### PR TITLE
Dispatch half type for some CPU functions

### PR DESCRIPTION
**Pull request by cmpute**
This PR added kHalf dispatching for some CPU functions. I add these dispatching to support [half precision in torch-scatter](https://github.com/rusty1s/pytorch_scatter/pull/137)

### CODE REVIEW

**Comment by mruberry** [comment]
Hey @cmpute, thanks for the PR! This looks pretty good but you'll need to update the tests so they (1) pass and (2) test all these new codepaths.

**Comment by cmpute** [comment]
@mruberry I think I have fixed the test and I rebased to the latest master

**Comment by mruberry** [comment]
> @mruberry I think I have fixed the test and I rebased to the latest master

Cool! But you also need to add new tests for these behaviors. Since this PR is attempting to change so many functions, what about creating a separate PR to start that just picks one of these functions or a few related functions? For example, mean, std_var, and prod? Then you can write a test or a few tests for those three functions that check how they handle bfloat16 and half inputs. The test should not only verify they no longer throw a runtime error, but also that their computed values are as numerically stable as expected.

**Comment by cmpute** [comment]
I added tests for operators that are modified (checked below). For other operators, they were not in the test before so I also skipped them (unchecked below)

- [x] lt
- [x] masked_fill
- [x] masked_select
- [ ] masked_select_serial
- [x] mean
- [x] std_var
- [x] prod
- [x] min
- [x] max
- [ ] softmax
- [ ] log_softmax
- [ ] softmax_lastdim
- [ ] log_softmax_lastdim
- [x] clamp
- [x] clamp_max
- [x] clamp_min
- [ ] unfolded2d_acc
- [ ] unfolded2d_copy
- [ ] unfolded3d_acc
- [ ] unfolded3d_copy

Noted that I skipped part of the test of `mean` operator since it will overflow when the array is too large: https://github.com/pytorch/pytorch/blob/c8045f581f34771abf68f5ce35e8aa64c4ee66df/test/test_torch.py#L610

This could be an separate issue? The failure case is like
```python
import torch
torch.tensor([1e38]*10).mean() # float32 overflow
torch.tensor([1e4]*10).half().mean() # float16 overflow
torch.tensor([1e4]*10).half().numpy().mean() # this won't overflow
torch.tensor([1e5]*10).half().numpy().mean() # this overflow
```
It seems like numpy has better way to deal with overflow in `mean` operator

File: test/test_torch.py [hunk 1]
```diff
@@ -19163,10 +19160,10 @@ def inner(self, device, dtype):
     ('chunk', '', _medium_2d, lambda t, d: [4], 1e-5, 1e-5, 1e-5, _types, _cpu_types, False),
     ('chunk', 'dim', _medium_2d, lambda t, d: [4, 1], 1e-5, 1e-5, 1e-5, _types, _cpu_types, False),
     ('chunk', 'neg_dim', _medium_2d, lambda t, d: [4, -2], 1e-5, 1e-5, 1e-5, _types, _cpu_types, False),
-    ('clamp', 'neg', _medium_2d, lambda t, d: [-1, 5], 1e-5, 1e-2, 1e-5, _signed_types, [torch.bfloat16]),
-    ('clamp', 'pos', _medium_2d, lambda t, d: [1, 5], 1e-5, 1e-2, 1e-5, _unsigned_types, [torch.bfloat16]),
-    ('clamp_min', '', _medium_2d, lambda t, d: [1], 1e-2, 1e-2, 1e-5, _types, [torch.bfloat16]),
-    ('clamp_max', '', _medium_2d, lambda t, d: [1], 1e-2, 1e-2, 1e-5, _types, [torch.bfloat16]),
+    ('clamp', 'neg', _medium_2d, lambda t, d: [-1, 5], 1e-5, 1e-2, 1e-5, _signed_types, [torch.bfloat16, torch.half]),
```
Comment (mruberry) [inline test/test_torch.py:19163]: Unfortunately adding torch.half here doesn't really test anything. This is, admittedly, super confusing and I'm working on changing the test suite to resolve this issue, so you can't test using these generated tests.

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp [hunk 2]
```diff
@@ -163,7 +163,7 @@ static void sign_kernel(TensorIterator& iter){
   if(iter.dtype() == ScalarType::Bool){
       cpu_kernel(iter, [=](bool x) -> bool { return x; });
   } else {
-    AT_DISPATCH_ALL_TYPES_AND2(kBFloat16, ScalarType::Half, iter.dtype(), "sign_cpu", [&]() {
+    AT_DISPATCH_ALL_TYPES_AND2(kBFloat16, kHalf, iter.dtype(), "sign_cpu", [&]() {
```
Comment (mruberry) [inline aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:166]: I'm actually creating a test suite that will make it easy to test this soon. It should probably wait for https://github.com/pytorch/pytorch/pull/41249.

File: test/test_torch.py (hunk 1 above)
Comment (mruberry) [reply test/test_torch.py:19163]: test_clamp in test_torch.py exists but doesn't instantiate dtype variants - maybe it could be updated?

File: aten/src/ATen/native/cpu/SoftMaxKernel.cpp [hunk 3]
```diff
@@ -224,25 +224,26 @@ struct vec_host_softmax_backward_lastdim {
 };
 
 static void softmax_lastdim_kernel_impl(Tensor& result, const Tensor& self) {
-  AT_DISPATCH_FLOATING_TYPES(self.scalar_type(), "softmax_lastdim_kernel_impl", [&] {
-    vec_host_softmax_lastdim<scalar_t, false>::apply(result, self);
+  AT_DISPATCH_FLOATING_TYPES_AND(kBFloat16,
```
Comment (mruberry) [inline aten/src/ATen/native/cpu/SoftMaxKernel.cpp:227]: We have some softmax testing in test_torch, but it looks like we'd need to update the softmax tests in test_nn.py to test this.

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp [hunk 4]
```diff
@@ -136,7 +136,7 @@ static void mean_kernel_impl(TensorIterator& iter) {
 }
 
 static void std_var_kernel_impl(TensorIterator &iter, bool unbiased, bool take_sqrt) {
-  AT_DISPATCH_FLOATING_TYPES_AND_HALF(iter.dtype(), "std_cpu", [&] {
+  AT_DISPATCH_FLOATING_TYPES_AND2(kBFloat16, kHalf, iter.dtype(), "std_cpu", [&] {
```
Comment (mruberry) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:139]: Tests of std and var exist in test_torch but would need to be updated to take a dtype variant

**Comment by mruberry** [comment]
CUDA mean doesn't seem to overflow in that case?

But that issue brings me back to my earlier point: I think we're going to have to break this PR up to make progress. There's just too much detail and testing required for each of these changes. The most complete change at this point looks to be `lt`, which is a straightforward modification to an existing test. How about starting with that one?

`clamp` is probably the next easiest, but it will require modifying `test_clamp` in test_torch.py.
